src/wfs/src/main/java/org/geoserver/wfs/json/GeoJSONBuilder.java

```diff
@@ -297,11 +301,17 @@ public static int getGeometryType(Geometry geometry) {
     public GeoJSONBuilder value(Object value) {
         if (value instanceof java.util.Date || value instanceof Calendar) {
             value = Converters.convert(value, String.class);
+        } else if (value instanceof Double) {
+            value = roundToNumberOfDecimalPlaces((Double)value);
         }
         super.value(value);
         return this;
     }
-    
+
+    private double roundToNumberOfDecimalPlaces(Double value) {
+        return new BigDecimal((Double)value).setScale(numDecimals, BigDecimal.ROUND_HALF_UP).doubleValue();		
```

Comment:
Ignore that, I realise now that -Math.ceil(-x * scale - 0.5) ==
Math.floor(x * scale + 0.5)!

On Mon, Jan 11, 2016 at 11:12 AM, Dean Povey <dean.povey@gmail.com> wrote:

> Okay, I did consider that, but wanted to keep the change minimal.
>
> I note that the code you reference does not correctly round negative
> values (or at least it is not consistent with any standard rounding scheme)
> it probably should be:
>
> x = (x < 0) ? -Math.ceil(-x * scale - 0.5) / scale : Math.floor(x * scale
> + 0.5) / scale;
>
> Which gives you the "Round Half Up" mode.
>
> Using BigDecimal was designed to handle the precision loss. You can find a
> small percentage of examples which fail when you round to 8 decimals, e.g:
>
> 6251.334203214999 rounded to 8 decimal places gives 6521.33420322 instead
> of 6251.33420321
>
> These will be rare.  There are some cases where even BigDecimal gives the
> wrong answer here so it is probably something you can live with.
>
> I will make a change to do this in primitive doubles as above if you are
> happy with those caveats.
>
>
> On Sat, Jan 9, 2016 at 8:12 PM, Andrea Aime <notifications@github.com>
> wrote:
>
>> In src/wfs/src/main/java/org/geoserver/wfs/json/GeoJSONBuilder.java
>> <https://github.com/geoserver/geoserver/pull/1392#discussion_r49262695>:
>>
>> >          }
>> >          super.value(value);
>> >          return this;
>> >      }
>> > -
>> > +
>> > +    private double roundToNumberOfDecimalPlaces(Double value) {
>> > +        return new BigDecimal((Double)value).setScale(numDecimals, BigDecimal.ROUND_HALF_UP).doubleValue();		
>>
>> Looking at this one again, this is a massive generator of garbage, by
>> allocating a large object as BigDecimal to format each and every ordinate.
>> While I haven't tried, I believe this is going to affect performance
>> significantly.
>> The GML2 output format got a nice speedup when we optimized its decimal
>> places handling, see here:
>>
>> https://github.com/geotools/geotools/blob/master/modules/library/main/src/main/java/org/geotools/gml/producer/CoordinateWriter.java#L295
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/geoserver/geoserver/pull/1392/files#r49262695>.
>>
>
>